dial_sync.go

```diff
@@ -127,14 +229,27 @@ func (ds *DialSync) getActiveDial(p peer.ID) *activeDial {
 // DialLock initiates a dial to the given peer if there are none in progress
 // then waits for the dial to that peer to complete.
 func (ds *DialSync) DialLock(ctx context.Context, p peer.ID) (*Conn, error) {
-	return ds.getActiveDial(p).wait(ctx)
-}
-
-// CancelDial cancels all in-progress dials to the given peer.
-func (ds *DialSync) CancelDial(p peer.ID) {
-	ds.dialsLk.Lock()
-	defer ds.dialsLk.Unlock()
-	if ad, ok := ds.dials[p]; ok {
-		ad.cancel()
+	var ad *activeDial
+
+startDial:
+	for {
+		ad = ds.getActiveDial(p)
+
+		// signal the start of dial
+		select {
+		case ad.dialch <- struct{}{}:
+			break startDial
+		case <-ad.waitch:
+			// we lost a race, we need to try again because the connection might not be what we want
+			ad.decref()
```

Comment:
Why are we dereferencing? Won't that cause the in-progress dial to be canceled if the original caller bails?

Reply:
I mean, we'll just restart, but that's pretty inefficient if we were half-way through opening a bunch of connections.

Reply:
Turns out, we only invoke this code if the active dial is currently "finishing". So we're not going to cancel anything.